### ISSUE TITLE
Unable to pull container with podman

### ISSUE THREAD

**Issue by fauust** [mvdan/sh#631]
Hi, this is not an issue.

I am moving most of my docker usage to podman (https://podman.io/).

For most of the cases it is relatively transparent but in the case of shfmt container, I am unable to pull the image:
```console
❯ podman pull mvdan/shfmt:v3.2.0
Trying to pull docker.io/mvdan/shfmt:v3.2.0...
  unsupported docker v2s2 media type: "application/vnd.oci.image.layer.v1.tar+gzip"
```

I am not sure where the incompatibility comes from (container itself so build process or docker registry). Maybe there is a workaround?

If you could provide a podman compatible container, that would be very appreciated. If I can help in that task, please let me know.

**Comment by mvdan** [comment]
I don't know. I build multi-architecture images with buildkit, so maybe that's too new/experimental for podman.

**Comment by fauust** [comment]
No idea, no urgency either, thanks for your response.
Here is an issue that may give information https://github.com/containers/buildah/issues/2266

**Comment by mvdan** [comment]
If someone wants to investigate this, I'll happily accept a pull request. Otherwise, I'll likely close this issue in a couple of weeks since I don't intend to investigate what needs to be done to support podman. Maintaining the necessary details to build multi-architecture images with buildkit (see CI) is already plenty of time taken by publishing to Docker.

**Comment by fauust** [comment]
I'll investigate this, can you confirm that the build/push is all done into github actions?

**Comment by mvdan** [comment]
Yes, the code is all [here](https://github.com/mvdan/sh/blob/d0927061ffa5d137ca779488192ee87254e68052/.github/workflows/test.yml#L39).

You can ignore the last step. It just updates the docker image's README with the README in this repo.

Similarly, the previous large step just tests that `docker run` works with the images as expected.

`Install Docker buildx` is also a lot of code to just... install a recent enough Docker version that has buildx. This was done when GitHub Actions used Ubuntu 18.04, which was too old. They've introduced 20.04 now, so maybe that step is not necessary anymore. I haven't dug into it yet.

The author of a lot of this code is @ArturKlauser, who might be kind enough to give tips or help in some other way.

**Comment by fauust** [comment]
So, what I have found out so far is that I am not able to reproduce the problem by building a very simple container locally (but I am able to reproduce the problem with shfmt container build locally):

- create a basic Dockerfile containing the following:

```Dockerfile
FROM alpine:3.12.1
RUN apk update && apk add bash
```
- start the local registry:

```console
docker run -it -p 127.0.0.1:5000:5000 registry:2
INFO[0000] Starting upload purge in 45m0s                go.version=go1.11.2 instance.id=422b93cc-ec6e-4fb2-9885-139fea8f0807 service=registry version=v2.7.1
INFO[0000] redis not configured                          go.version=go1.11.2 instance.id=422b93cc-ec6e-4fb2-9885-139fea8f0807 service=registry version=v2.7.1
INFO[0000] using inmemory blob descriptor cache          go.version=go1.11.2 instance.id=422b93cc-ec6e-4fb2-9885-139fea8f0807 service=registry version=v2.7.1
INFO[0000] listening on [::]:5000                        go.version=go1.11.2 instance.id=422b93cc-ec6e-4fb2-9885-139fea8f0807 service=registry version=v2.7.1
```

- create/use the buildx builder:

```console
docker buildx create --use --name build --node build --driver-opt network=host
```

- build, push container:

```console
docker buildx build --push -t localhost:5000/test:alpine .
```

- pull and test container with docker:

```console
docker run -it test:alpine bash -c uname
Linux
```

- pull and test container with podman:

Make sure that you have the following in `~/.config/containers/registries.conf`:

```text
[registries.insecure]
registries = ['localhost']
```

```console
podman run -it test:alpine bash -c uname
Linux
```

Whereas if I follow the same steps and build/push the shfmt container I can reproduce the problem:

- clone - build - push

```console
git clone https://github.com/mvdan/sh shfmt
cd shfmt
docker buildx build --push -t localhost:5000/test:shfmt -f cmd/shfmt/Dockerfile .
```

- run with docker

```console
docker run -it localhost:5000/test:shfmt --version
v3.2.0-16-g9becc72e
```

- run with podman

```console
podman run -it localhost:5000/test:shfmt --version
Trying to pull localhost:5000/test:shfmt...
  unsupported docker v2s2 media type: "application/vnd.oci.image.layer.v1.tar+gzip"
Error: unable to pull localhost:5000/test:shfmt: Error initializing image from source docker://localhost:5000/test:shfmt: unsupported docker v2s2 media type: "application/vnd.oci.image.layer.v1.tar+gzip"
```

So, I guess that something in the build process (maybe the fact that it is using multiple containers to build the final container) is causing the problem... I'll investigate further in the next days. Also I may open an issue there https://github.com/docker/buildx/issues.

**Comment by fauust** [comment]
> `Install Docker buildx` is also a lot of code to just... install a recent enough Docker version that has buildx. This was done when GitHub Actions used Ubuntu 18.04, which was too old. They've introduced 20.04 now, so maybe that step is not necessary anymore. I haven't dug into it yet.

See: https://github.com/mvdan/sh/pull/638

**Comment by ArturKlauser** [comment]
Thanks for updating this. I've added some comments on #638.

**Comment by fauust** [comment]
Thanks for your comments @ArturKlauser, very useful!
Do you have any idea on how to resolve this issue BTW? Or maybe who to ping, maybe https://github.com/docker/buildx/issues?

**Comment by ArturKlauser** [comment]
Unfortunately I don't. I've never used podman.

**Comment by fauust** [comment]
@ArturKlauser, as you suggested in #638, I am [preparing](https://github.com/fauust/sh/blob/build-and-test-locally-first/.github/workflows/test.yml) the new workflow (testing images against local registry before pushing to DockerHub). Everything seems to work perfectly even if I may ping you when it's ready because I am not very sure that it's fully optimized (we should not build containers twice and make sure to use local registry cache I guess).

I have read carefully your excellent [post](https://nexus.eddiesinentropy.net/2020/01/12/Building-Multi-architecture-Docker-Images-With-Buildx/) and there is still one thing that I do not understand about the actual workflow: What is the purpose of creating and targeting `-alpine` images?

The good (odd!) news is that I am able to use the containers that I have generated [yesterday](https://hub.docker.com/layers/fauust/shfmt/latest/images/sha256-f1b09131944ba05e5af21a56d78a589bac76e46c7013f9df15059bd79083c913?context=explore) with the new workflow:
```console
❯ podman run -it fauust/shfmt@sha256:f1b09131944ba05e5af21a56d78a589bac76e46c7013f9df15059bd79083c913 -h
Trying to pull docker.io/fauust/shfmt@sha256:f1b09131944ba05e5af21a56d78a589bac76e46c7013f9df15059bd79083c913...
Getting image source signatures
Copying blob 879400b486fb done
Copying config 741efb2910 done
Writing manifest to image destination
Storing signatures
usage: shfmt [flags] [path ...]
...
For more information, see 'man shfmt' and https://github.com/mvdan/sh.
```

And the other good (and also strange) news is that podman is able to work with actual `-alpine` images (I did not test this before!):
```console
❯ podman run -it mvdan/shfmt:latest-alpine -h
Trying to pull docker.io/mvdan/shfmt:latest-alpine...
Getting image source signatures
Copying blob ea3f670d8fa6 done
Copying blob 141232587160 done
Copying blob 188c0c94c7c5 done
Copying config 1638c8d0ca done
Writing manifest to image destination
Storing signatures
usage: shfmt [flags] [path ...]
...
For more information, see 'man shfmt' and https://github.com/mvdan/sh.
```

**Comment by ArturKlauser** [comment]
That looks pretty good. Does the caching work, i.e. is it reusing the image that was built and pushed to the local repository when it pushes to dockerhub?

> What is the purpose of creating and targeting `-alpine` images?

I've seen on several projects that both naked and -alpine docker images are offered. I've never needed to use the -alpine image, but there are CI systems which apparently need a full image for every step. I think [Circle CI](https://circleci.com/docs/2.0/executor-types/) is one of them. @mvdan can probably tell you more.

**Comment by fauust** [comment]
> That looks pretty good. Does the caching work, i.e. is it reusing the image that was built and pushed to the local repository when it pushes to dockerhub?

Well as said above, that's exactly where I will appreciate your expert review :-)
Looking at times, I guess that yes but maybe not everything is reused from cache, see: https://github.com/fauust/sh/runs/1470499748?check_suite_focus=true
![screenshot_20201130_181112](https://user-images.githubusercontent.com/2197555/100641749-2e208380-3338-11eb-9ba4-f166ee13222d.png)

**Comment by ArturKlauser** [comment]
I looked through the logs and agree, it seems that a lot gets reused, but not everything - strange.
I've never tried to use --cache-from in a docker build, so that's new territory for me too.
With regard to debugging, I found it easiest to only try it with a single --platform, since otherwise the intermingled output between the different platform builds becomes hard to follow.

**Comment by fauust** [comment]
> I looked through the logs and agree, it seems that a lot gets reused, but not everything - strange.
> I've never tried to use --cache-from in a docker build, so that's new territory for me too.

I guess that this is documented here https://github.com/docker/buildx#--cache-fromnametypetypekeyvalue maybe we should play with the cache `mode`. The strange thing is that I would have spontaneously thought that having the everything build locally would not require us to play with any specific caching option and that `docker build` would automatically reuse the maximum from the local registry, but apparently that's not the case...

> With regard to debugging, I found it easiest to only try it with a single --platform, since otherwise the intermingled output between the different platform builds becomes hard to follow.

That's a good idea, I will use it once #638 is merged in the preparation of the new build workflow.

**Comment by ArturKlauser** [comment]
I wonder if a more direct way of copying the built image from the local registry to docker hub would work better. E.g. https://github.com/containers/skopeo or https://github.com/mdlavin/copy-docker-image seem to be able to do that. But I haven't tried it, so I don't know if it works with multi-platform images.

**Comment by fauust** [comment]
> I wonder if a more direct way of copying the built image from the local registry to docker hub would work better. E.g. https://github.com/containers/skopeo or https://github.com/mdlavin/copy-docker-image seem to be able to do that. But I haven't tried it, so I don't know if it works with multi-platform images.

I will check this, thanks for the pointers.

**Comment by fauust** [comment]
With #638 merged, the containers are now working with podman, I did not understand why though.

```console
❯ podman run -it mvdan/shfmt:latest -h
Trying to pull docker.io/mvdan/shfmt:latest...
Getting image source signatures
Copying blob c039d061f8f2 done
Copying config 569301377b done
Writing manifest to image destination
Storing signatures
usage: shfmt [flags] [path ...]
...
❯ podman run -it mvdan/shfmt:latest-alpine -h
Trying to pull docker.io/mvdan/shfmt:latest-alpine...
Getting image source signatures
Copying blob 188c0c94c7c5 done
Copying blob a3d49ffea543 done
Copying blob 04170ef31153 done
Copying config 9a56227044 done
Writing manifest to image destination
Storing signatures
usage: shfmt [flags] [path ...]
```

Closing this issue.

**Comment by mvdan** [comment]
Good to hear, thanks for following up!
